janitor/utils.py

```diff
@@ -405,3 +392,229 @@ def _wrapped(x, *args, **kwargs):
+                raise TypeError(
+                    "`expand_grid` works with only vector and matrix arrays")
+        # process series
+        if isinstance(value, pd.Series):
+            if value.empty:
+                raise ValueError("passed Series cannot be empty")
+            if not isinstance(value.index, pd.MultiIndex):
+                # this section checks if the Series has a name or not
+                # and uses that information to create a new column name
+                # for the resulting Dataframe
+                if value.name:
+                    value = value.to_frame(name=f"{key}_{value.name}")
+                    dfs.append(value)
+                else:
+                    value = value.to_frame(name=f"{key}")
+                    dfs.append(value)
+            else:
+                raise TypeError(
+                    "`expand_grid` does not work with pd.MultiIndex")
+        # process dataframe
+        if isinstance(value, pd.DataFrame):
+            if value.empty:
+                raise ValueError("passed DataFrame cannot be empty")
+            if not (isinstance(value.index, pd.MultiIndex)
+                    or isinstance(value.columns, pd.MultiIndex)):
+                # add key to dataframe columns
+                value = value.add_prefix(f"{key}_")
+                dfs.append(value)
+            else:
+                raise TypeError(
+                    "`expand_grid` does not work with pd.MultiIndex")
+        # process lists
+        if isinstance(value, list):
+            if not value:
+                raise ValueError("passed Sequence cannot be empty")
+            if np.array(value).ndim == 1:
+                checklist = (type(None), str, int, float, bool)
+                instance_check_type = (isinstance(internal, checklist)
+                                       for internal in value)
+                if all(instance_check_type):
+                    dicts.update({key: value})
+                else:
+                    raise ValueError("values in iterable must be scalar")
+            elif np.array(value).ndim == 2:
+                value = pd.DataFrame(value).add_prefix(f"{key}_")
+                dfs.append(value)
+            else:
+                raise ValueError("Sequence's dimension should be 1d or 2d")
+
+    return dfs, dicts
+
+
+def _grid_computation_dict(dicts: Dict) -> pd.DataFrame:
+    """
+    Function used within the expand_grid function,
+    to compute dataframe from values that are not dataframes/arrays/series.
+    These values are collected into a dictionary,
+    and processed with numpy meshgrid.
+
+    Numpy's meshgrid is faster than itertools' product,
+    and when converting to a dataframe, is fast as well.
+    Structured arrays are used here, to ensure the datatypes are preserved.
+    """
+    # if there is only name value pair in the dictionary
+    if len(dicts) == 1:
+        key = list(dicts.keys())[0]
+        value = list(dicts.values())[0]
```

Comment:
The linter is definitely getting this wrong.

```suggestion
        value = list(dicts.values())[0]  # noqa: PD011
```